File: maze.rs
// Maze generator in Rust.
// 2014-01-11 <> 20151208
// Joe Wingbermuehle

extern crate rand;

use rand::Rng;
use std::vec::Vec;

enum Element {
    Wall,
    Passage,
}

struct Maze {
    width:      usize,
    height:     usize,
    data:       Vec<Element>,
}

impl Maze {

    // Create an empty maze structure.
    fn init(width: usize, height: usize) -> Maze {
        let mut maze = Maze {
            width:  width,
            height: height,
            data:   Vec::new(),
        };
        for y in 0 .. height {
            for x in 0 .. width {
                let value =
                    if y == 0 || y == height - 1 {
                        Element::Passage
                    } else if x == 0 || x == width - 1 {
                        Element::Passage
                    } else {
                        Element::Wall
                    };
                maze.data.push(value);
            }
        }
        maze
    }

    // Show the maze.
    fn show(&self) {
        for y in 0 .. self.height {
            for x in 0 .. self.width {
                match self.data[y * self.width + x] {
                   Element::Wall => print!("[]"),
                   Element::Passage => print!("  "),
                }
            }
            println!("")
        }
    }

    fn is_wall(&self, x: isize, y: isize) -> bool {
        let (ux, uy) = (x as usize, y as usize);
        match self.data[(uy * self.width + ux) as usize] {
            Element::Wall => true,
            Element::Passage => false,
        }
    }

    // Carve the maze starting at x, y.
    fn carve<R: rand::Rng>(&mut self, rng: &mut R, x: usize, y: usize) {
        let xdirs = [ 1, -1, 0, 0 ];
        let ydirs = [ 0, 0, 1, -1 ];
        self.data[y * self.width + x] = Element::Passage;
        let d = rng.gen::<usize>() % 4;
        for i in 0 .. 4 {
            let dx: isize = xdirs[(d + i) % 4];
            let dy: isize = ydirs[(d + i) % 4];
            let x2 = (x as isize) + dx;
            let y2 = (y as isize) + dy;
            if self.is_wall(x2, y2) {
                let nx = x2 + dx;
                let ny = y2 + dy;
                if self.is_wall(nx, ny) {
                    let index = (y2 as usize) * self.width + (x2 as usize);
                    self.data[index] = Element::Passage;
                    self.carve(rng, nx as usize, ny as usize);
                }
            }
        }

    }

    // Generate a maze.
    fn generate(width: usize, height: usize) -> Maze {
        let mut maze = Maze::init(width, height);
        let mut rng = rand::thread_rng();
        maze.carve(&mut rng, 2, 2);
        maze.data[1 * width + 2] = Element::Passage;
        maze.data[(height - 2) * width + (width - 3)] = Element::Passage;
        maze
    }

}

fn main() {
    let width = 39;     // Maze width; must be odd.
    let height = 23;    // Maze height; must be odd.
    let maze = Maze::generate(width, height);
    maze.show();
}

File: maze.swift
// Maze generator in Swift
// 2014-12-11 - Joe Wingbermuehle
// 2018-02-20 - Sarp Başaraner - Update for Swift 4
// Build with:
//       xcrun -sdk macosx swift maze.swift 

import Foundation

class Maze {

    enum Cell {
        case Space, Wall
    }

    var data: [[Cell]] = []

    // Generate a random maze.
    init(width: Int, height: Int) {
        for _ in 0 ..< height {
            data.append([Cell](repeating: Cell.Wall, count: width))
        }
        for i in 0 ..< width {
            data[0][i] = Cell.Space
            data[height - 1][i] = Cell.Space
        }
        for i in 0 ..< height {
            data[i][0] = Cell.Space
            data[i][width - 1] = Cell.Space
        }
        data[2][2] = Cell.Space
        self.carve(x: 2, y: 2)
        data[1][2] = Cell.Space
        data[height - 2][width - 3] = Cell.Space
    }

    // Carve starting at x, y.
    func carve(x: Int, y: Int) {
        let upx = [1, -1, 0, 0]
        let upy = [0, 0, 1, -1]
        var dir = Int(arc4random_uniform(4))
        var count = 0
        while count < 4 {
            let x1 = x + upx[dir]
            let y1 = y + upy[dir]
            let x2 = x1 + upx[dir]
            let y2 = y1 + upy[dir]
            if data[y1][x1] == Cell.Wall && data[y2][x2] == Cell.Wall {
                data[y1][x1] = Cell.Space
                data[y2][x2] = Cell.Space
                carve(x: x2, y: y2)
            } else {
                dir = (dir + 1) % 4
                count += 1
            }
        }
    }

    // Show the maze.
    func show() {
        for row in data {
            for cell in row {
                if cell == Cell.Space {
                    print("  ", separator: "", terminator: "")
                } else {
                    print("[]", separator: "", terminator: "")
                }
            }
            print("")
        }
    }

}

let maze = Maze(width: 39, height: 23)
maze.show()
